driver-core/src/main/java/com/datastax/driver/core/policies/TokenAwarePolicy.java

```diff
@@ -58,6 +62,82 @@
  */
 public class TokenAwarePolicy implements ChainableLoadBalancingPolicy {
 
+  public static String getTokenRange(Statement statement) {
+    if (statement == null) return null;
+    String query = null;
+    if (statement instanceof BoundStatement) {
+      query = ((BoundStatement) statement).preparedStatement().getQueryString();
+    } else {
+      query = statement.toString();
+    }
+    StringTokenizer st = new StringTokenizer(query);
+    try {
+      if (st.nextToken().equalsIgnoreCase("select")) {
+        while (st.hasMoreElements() && !st.nextToken().equalsIgnoreCase("where")) ;
+        int valueCounter = 0;
+        boolean hasToken = false;
+        boolean appending = false;
+        List<String> tokens = new ArrayList<>();
+        while (st.hasMoreElements()) {
```

Comment:
below can be done more effectively using a lexer (see jflex and similar)

Reply:
resp. there must be a query parser, so we shall use it instead of string based parsing, which is very noneffective and slows down this method (if it's called often, then this is a problem)

Reply:
Hi, Lubos. I first developed this outside the driver code and tested it a number of ways. My first approach was to parse it with a regular expression, but it became way cryptic to be easily understood by anyone not familiar with the code. Also, the initial performance tests that I conducted showed no improvements at all with this approach. I know that parsing a String looks ugly, but it's not as inneficient as it might look. But I'd be interested in knowing better alternatives, if you have any suggestion.